srv/modules/runners/validate.py

```diff
@@ -1054,6 +1074,150 @@ def report(self):
+                                                         suffix=self.suffix))
+        self.map = self.load_map()
+        self.issues = []
+
+    def load_map(self):
+        """
+        Loads k:v map from disk
+
+        Returns:
+            YAML map
+        Raises:
+            YAMLError
+        """
+        with open(self.map_file, 'r') as _fd:
+            try:
+                return yaml.load(_fd)
+            except yaml.YAMLError:
+                log.error('Could not read {}'.format(self.map_file))
+
+    def read_lines(self, filename):
+        """ Reads lines from an open file and returns a generator
+        Args:
+            fn (str): filename
+        Yields:
+            str: line from file
+        """
+        with open(filename, 'r') as _fd:
+            for line in _fd.readlines():
+                yield line
+
+    def extract_k_v(self, line):
+        """ Extracts key and value from line
+        Args:
+            line (str): line from file
+        Returns:
+            tuple: stripped key and value
+        """
+        _key, _value = line.split('=')
+        return _key.strip(), _value.strip()
+
+    def check_line(self, line):
+        """ Checks a line for deprecated keys/values
+        Args:
+            line (str): line from file
+        Returns:
+            DeprecatedConf: instance of obj
+        """
+        _key, _value = self.extract_k_v(line)
+        return self.compare_k_v_to_map(_key, _value)
+
+    def compare_k_v_to_map(self, key, value):
+        """ Compares k:v against a map of k:v that are know to be deprecated
+        Args:
+            k (str): key from config
+            v (str): value from config
+        Returns:
+            DeprecatedConf: instance of obj or None
+        """
+        obj = None
+        for release, kv_map in self.map.items():
+            if key not in kv_map:
+                continue
+            if isinstance(kv_map[key], list):
+                obj = DeprecatedConf(key=key,
+                                     release=release)
+                for depr_val in kv_map[key]:
+                    if value == depr_val:
+                        obj.add_value(depr_val)
+            if isinstance(kv_map[key], str):
+                if kv_map[key] == 'any':
+                    obj = DeprecatedConf(key=key,
+                                         release=release,
+                                         values=[])
+                if kv_map[key] == value:
+                    obj = DeprecatedConf(key=key,
+                                         release=release,
+                                         values=[value])
+        return obj
+
+    def run(self):
+        """
+        Returns:
+            list: contains objects of DeprecatedConf
+        """
+        for filename in self.files:
+            for line in self.read_lines(filename):
+                conf_object = self.check_line(line)
```

Comment:
Python includes a configparser class that knows how to read .ini style config files, which we've used in a couple of places already, which would save adding functions to extract key-value data. Something like this:

````
conf = configparser.RawConfigParser()
with open(filename) as lines:
    conf.read_string('\n'.join(line.strip() for line in lines))
for section in conf.sections():
    if conf.has_option(section, "key to check goes here"):
        # found deprecated option, warn about it
````

Reply:
> The ConfigParser needs sections in the .conf(.ini) files, which we never agreed to use in any of (/srv/salt/ceph/configuration/files/ceph.conf.d/*.conf).

Ah. That difference didn't occur to me :-/

> Alternatively we could you python-configobj which seems to be able to load a .ini file without sections.

Looks like we've already got "Requires: python3-configobj" in the spec file, so, sure, go for it :-)

Reply:
Pushed a commit to address your suggestions.

* underscores in keys will be normalized
* configobj is being used
* adapted tests and added tests for normalization method